torch/csrc/autograd/profiler.cpp

```diff
@@ -156,98 +281,126 @@ void enableProfiler(ProfilerConfig config) {
+        if (profiler_nested_depth_ == 0) {
+          pushProfilingCallbacks();
+        }
+        profiler_nested_depth_++;
+      }
+      if (!to_push) {
+        --profiler_nested_depth_;
+        if (profiler_nested_depth_ == 0) {
+          removeProfilingCallbacks();
+        }
+      }
+    }
+  );
+  return true;
+}();
+
+const int kCUDAWarmupStart = 5;
+
+} // namespace
+
+void registerCUDAMethods(CUDAStubs* stubs) {
+  cuda_stubs = stubs;
+}
+
+ProfilerConfig::~ProfilerConfig() = default;
+
+bool profilerEnabled() {
+  const auto& state = at::ThreadLocalDebugInfo::get(at::DebugInfoKind::PROFILER_STATE);
+  auto state_ptr = dynamic_cast<ProfilerThreadLocalState*>(state.get());
+  return state_ptr && state_ptr->config().state != ProfilerState::Disabled;
+}
+
+void enableProfiler(const ProfilerConfig& new_config) {
+  TORCH_CHECK(new_config.state != ProfilerState::NVTX || cuda_stubs->enabled(),
+    "Can't use NVTX profiler - PyTorch was compiled without CUDA");
+
+  auto state = std::make_shared<ProfilerThreadLocalState>(new_config);
+  at::ThreadLocalDebugInfo::push(at::DebugInfoKind::PROFILER_STATE, state);
+
+  if (profiler_nested_depth_ == 0) {
+    pushProfilingCallbacks(new_config.report_input_shapes);
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Profiler, true);
+  }
+  ++profiler_nested_depth_;
+
+  if (new_config.state == ProfilerState::CUDA) {
     // event recording appears to have some startup overhead, so we need to
     // to generate some dummy events first before recording synchronization events
-    for(int i = 0; i < 5; i++) {
-      cuda_stubs->onEachDevice([](int d) {
-          mark("__cuda_startup");
+    for (int i = 0; i < kCUDAWarmupStart; i++) {
+      cuda_stubs->onEachDevice([state](int /* unused */) {
+          state->mark("__cuda_startup");
           cuda_stubs->synchronize();
       });
     }
 
     // cuda events must be on the same device, so we need a start event recorded
     // for each gpu. we then use this event to synchronize time on the GPU
     // with the CPU clock.
-    cuda_stubs->onEachDevice([](int d) {
-        mark("__cuda_start_event");
+    cuda_stubs->onEachDevice([state](int d) {
+        state->mark("__cuda_start_event");
     });
   }
-  mark("__start_profile", false);
+  state->mark("__start_profile", false);
 }
 
 thread_event_lists disableProfiler() {
-  if (state == ProfilerState::Disabled) {
-    throw std::runtime_error("can't disable profiler when it's not running");
+  auto state = at::ThreadLocalDebugInfo::pop(at::DebugInfoKind::PROFILER_STATE);
+  auto state_ptr = dynamic_cast<ProfilerThreadLocalState*>(state.get());
+  TORCH_CHECK(state_ptr && state_ptr->config().state != ProfilerState::Disabled,
+      "Can't disable profiler when it's not running");
+
+  --profiler_nested_depth_;
+  if (profiler_nested_depth_ == 0) {
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Profiler, false);
+    removeProfilingCallbacks();
```

Comment:
Do we need to do this if we move the thread local callbacks in `ProfilerThreadLocalState`? Seems like `at::ThreadLocalDebugInfo::pop(at::DebugInfoKind::PROFILER_STATE)` would do the necessary cleanup if we do that.

Reply:
answered above